### PR TITLE
move datagram to separate crate

### PR DESCRIPTION
**Pull request by Ruben2424**
The goal of This PR is to separate the [h3-datagram](https://datatracker.ietf.org/doc/html/rfc9297) logic from the `h3` crate providing a new crate `h3-datagram`.

### CODE REVIEW

File: h3-quinn/Cargo.toml [hunk 1]
```diff
@@ -21,6 +21,7 @@ quinn = { version = "0.11", default-features = false, features = [
 tokio-util = { version = "0.7.9" }
 futures = { version = "0.3.28" }
 tokio = { version = "1", features = ["io-util"], default-features = false }
+h3-datagram = { path = "../h3-datagram" }
```
Comment (seanmonstar) [inline h3-quinn/Cargo.toml:24]: Wondering, should it be optional? Is there enough code in here that people would benefit if they didn't need to compile it?

**Review by seanmonstar — COMMENTED** [review]
Generally looks great, good work!

Might be good if @darioalessandro or @ten3roberts could look, to see if they can still use it.

File: h3-quinn/Cargo.toml (hunk 1 above)
Comment (Ruben2424) [reply h3-quinn/Cargo.toml]: > Wondering, should it be optional? Is there enough code in here that people would benefit if they didn't need to compile it?

Changed it.

**Comment by Ruben2424** [comment]
> Generally looks great, good work!
> 
> Might be good if @darioalessandro or @ten3roberts could look, to see if they can still use it.

@seanmonstar i think you can continue with the review.
your comment got a :+1: . Additionally I tested against https://webtransport.day/ and it still worked.

**Comment by ten3roberts** [comment]
I'm no longer using WT as much, but I had a look and it looks great.

Nothing suspicious in the code, and good use of pub/pub(crate) and doc-hidden where needed :+1:

**Review by seanmonstar — APPROVED** [review]

**Pull request merged**